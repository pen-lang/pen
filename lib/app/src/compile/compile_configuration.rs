pub struct CompileConfiguration {
    pub heap: HeapConfiguration,
    pub list_type: ListTypeConfiguration,
}

pub type HeapConfiguration = fmm_llvm::HeapConfiguration;
pub type ListTypeConfiguration = lang::hir_mir::ListTypeConfiguration;
