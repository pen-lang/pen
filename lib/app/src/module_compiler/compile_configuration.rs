pub struct CompileConfiguration {
    pub instruction: InstructionConfiguration,
    pub list_type: ListTypeConfiguration,
    pub string_type: StringTypeConfiguration,
    pub error_type: ErrorTypeConfiguration,
}

pub type InstructionConfiguration = fmm_llvm::InstructionConfiguration;
pub type ListTypeConfiguration = hir_mir::ListTypeConfiguration;
pub type StringTypeConfiguration = hir_mir::StringTypeConfiguration;
pub type ErrorTypeConfiguration = hir_mir::ErrorTypeConfiguration;
