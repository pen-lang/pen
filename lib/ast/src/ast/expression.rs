use super::{
    BinaryOperation, ByteString, Call, If, IfList, IfMap, IfType, Lambda, List, ListComprehension,
    Map, Number, Record, RecordDeconstruction, UnaryOperation, Variable,
};
use position::Position;

#[derive(Clone, Debug, PartialEq)]
pub enum Expression {
    BinaryOperation(BinaryOperation),
    Call(Call),
    If(If),
    IfList(IfList),
    IfMap(IfMap),
    IfType(IfType),
    Lambda(Lambda),
    List(List),
    ListComprehension(ListComprehension),
    Map(Map),
    Number(Number),
    Record(Record),
    RecordDeconstruction(RecordDeconstruction),
    String(ByteString),
    UnaryOperation(UnaryOperation),
    Variable(Variable),
}

impl Expression {
    pub fn position(&self) -> &Position {
        match self {
            Self::BinaryOperation(operation) => operation.position(),
            Self::Call(call) => call.position(),
            Self::If(if_) => if_.position(),
            Self::IfList(if_) => if_.position(),
            Self::IfMap(if_) => if_.position(),
            Self::IfType(if_) => if_.position(),
            Self::Lambda(lambda) => lambda.position(),
            Self::List(list) => list.position(),
            Self::ListComprehension(comprehension) => comprehension.position(),
            Self::Map(map) => map.position(),
            Self::Number(number) => number.position(),
            Self::Record(record) => record.position(),
            Self::RecordDeconstruction(operation) => operation.position(),
            Self::String(string) => string.position(),
            Self::UnaryOperation(operation) => operation.position(),
            Self::Variable(variable) => variable.position(),
        }
    }
}

impl From<BinaryOperation> for Expression {
    fn from(operation: BinaryOperation) -> Self {
        Self::BinaryOperation(operation)
    }
}

impl From<Call> for Expression {
    fn from(call: Call) -> Self {
        Self::Call(call)
    }
}

impl From<IfType> for Expression {
    fn from(if_: IfType) -> Self {
        Self::IfType(if_)
    }
}

impl From<ByteString> for Expression {
    fn from(string: ByteString) -> Self {
        Self::String(string)
    }
}

impl From<If> for Expression {
    fn from(if_: If) -> Self {
        Self::If(if_)
    }
}

impl From<IfList> for Expression {
    fn from(if_: IfList) -> Self {
        Self::IfList(if_)
    }
}

impl From<IfMap> for Expression {
    fn from(if_: IfMap) -> Self {
        Self::IfMap(if_)
    }
}

impl From<Lambda> for Expression {
    fn from(lambda: Lambda) -> Self {
        Self::Lambda(lambda)
    }
}

impl From<List> for Expression {
    fn from(list: List) -> Self {
        Self::List(list)
    }
}

impl From<ListComprehension> for Expression {
    fn from(comprehension: ListComprehension) -> Self {
        Self::ListComprehension(comprehension)
    }
}

impl From<Map> for Expression {
    fn from(map: Map) -> Self {
        Self::Map(map)
    }
}

impl From<Number> for Expression {
    fn from(number: Number) -> Self {
        Self::Number(number)
    }
}

impl From<Record> for Expression {
    fn from(record: Record) -> Self {
        Self::Record(record)
    }
}

impl From<RecordDeconstruction> for Expression {
    fn from(operation: RecordDeconstruction) -> Self {
        Self::RecordDeconstruction(operation)
    }
}

impl From<UnaryOperation> for Expression {
    fn from(operation: UnaryOperation) -> Self {
        Self::UnaryOperation(operation)
    }
}

impl From<Variable> for Expression {
    fn from(variable: Variable) -> Self {
        Self::Variable(variable)
    }
}
