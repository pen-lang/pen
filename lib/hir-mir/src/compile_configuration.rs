use super::{
    error_type_configuration::ErrorTypeConfiguration,
    list_type_configuration::ListTypeConfiguration, map_type_configuration::MapTypeConfiguration,
    string_type_configuration::StringTypeConfiguration,
};
#[cfg(test)]
use super::{
    error_type_configuration::ERROR_TYPE_CONFIGURATION,
    list_type_configuration::LIST_TYPE_CONFIGURATION,
    map_type_configuration::MAP_TYPE_CONFIGURATION,
    string_type_configuration::STRING_TYPE_CONFIGURATION,
};
#[cfg(test)]
use once_cell::sync::Lazy;

#[derive(Clone, Debug)]
pub struct CompileConfiguration {
    pub list_type: ListTypeConfiguration,
    pub map_type: MapTypeConfiguration,
    pub string_type: StringTypeConfiguration,
    pub error_type: ErrorTypeConfiguration,
    pub spawn_function_name: String,
    pub debug_function_name: String,
}

#[cfg(test)]
pub static COMPILE_CONFIGURATION: Lazy<CompileConfiguration> = Lazy::new(|| CompileConfiguration {
    list_type: LIST_TYPE_CONFIGURATION.clone(),
    map_type: MAP_TYPE_CONFIGURATION.clone(),
    string_type: STRING_TYPE_CONFIGURATION.clone(),
    error_type: ERROR_TYPE_CONFIGURATION.clone(),
    spawn_function_name: "spawn".into(),
    debug_function_name: "debug".into(),
});
