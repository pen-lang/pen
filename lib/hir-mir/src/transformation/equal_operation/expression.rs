use super::operation;
use crate::{context::CompileContext, error::CompileError};
use hir::{analysis::AnalysisError, ir::*};

pub fn transform(
    context: &CompileContext,
    operation: &EqualityOperation,
) -> Result<Expression, CompileError> {
    Ok(if operation.operator() == EqualityOperator::Equal {
        operation::transform(
            context,
            operation
                .type_()
                .ok_or_else(|| AnalysisError::TypeNotInferred(operation.position().clone()))?,
            operation.lhs(),
            operation.rhs(),
            operation.position(),
        )?
    } else {
        operation.clone().into()
    })
}
