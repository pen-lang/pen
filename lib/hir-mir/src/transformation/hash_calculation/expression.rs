use super::function;
use crate::{
    context::CompileContext,
    transformation::{collection_type, map_context, record_type_information},
    CompileError,
};
use hir::{
    analysis::{
        type_comparability_checker, type_resolver, union_type_member_calculator, AnalysisError,
    },
    ir::*,
    types::{self, Type},
};
use position::Position;

const NONE_HASH: f64 = 0.0;
const BOOLEAN_TRUE_HASH: f64 = 1.0;
const BOOLEAN_FALSE_HASH: f64 = 2.0;

pub fn transform(
    context: &CompileContext,
    value: &Expression,
    type_: &Type,
    position: &Position,
) -> Result<Expression, CompileError> {
    let configuration = context.configuration()?;

    Ok(match type_ {
        Type::Boolean(_) => If::new(
            value.clone(),
            Number::new(BOOLEAN_TRUE_HASH, position.clone()),
            Number::new(BOOLEAN_FALSE_HASH, position.clone()),
            position.clone(),
        )
        .into(),
        Type::List(list_type) => Call::new(
            Some(
                types::Function::new(
                    vec![
                        compile_any_function_type(position).into(),
                        collection_type::transform_list(context, position)?,
                    ],
                    types::Number::new(position.clone()),
                    position.clone(),
                )
                .into(),
            ),
            Variable::new(
                &configuration.map_type.hash.list_hash_function_name,
                position.clone(),
            ),
            vec![
                function::transform(context, list_type.element())?,
                value.clone(),
            ],
            position.clone(),
        )
        .into(),
        Type::Map(map_type) => Call::new(
            Some(
                types::Function::new(
                    vec![
                        collection_type::transform_map_context(context, position)?,
                        collection_type::transform_map(context, position)?,
                    ],
                    types::Number::new(position.clone()),
                    position.clone(),
                )
                .into(),
            ),
            Variable::new(
                &configuration.map_type.hash.map_hash_function_name,
                position.clone(),
            ),
            vec![
                map_context::expression::transform(context, map_type)?,
                value.clone(),
            ],
            position.clone(),
        )
        .into(),
        Type::None(_) => Number::new(NONE_HASH, position.clone()).into(),
        Type::Number(_) => compile_concrete_hash_function_call(
            &configuration.map_type.hash.number_hash_function_name,
            value,
            type_,
            position,
        ),
        Type::Record(record_type) => {
            if !type_comparability_checker::check(type_, context.types(), context.records())? {
                return Err(CompileError::InvalidRecordEqualOperation(position.clone()));
            }

            compile_concrete_hash_function_call(
                record_type_information::compile_hash_function_name(record_type),
                value,
                type_,
                position,
            )
        }
        Type::String(_) => compile_concrete_hash_function_call(
            &configuration.map_type.hash.string_hash_function_name,
            value,
            type_,
            position,
        ),
        Type::Union(_) => {
            const VALUE_NAME: &str = "$x";
            let member_types = union_type_member_calculator::calculate(type_, context.types())?;

            IfType::new(
                VALUE_NAME,
                value.clone(),
                member_types
                    .iter()
                    .map(|type_| {
                        Ok(IfTypeBranch::new(
                            type_.clone(),
                            transform(
                                context,
                                &Variable::new(VALUE_NAME, position.clone()).into(),
                                type_,
                                position,
                            )?,
                        ))
                    })
                    .collect::<Result<_, CompileError>>()?,
                None,
                position.clone(),
            )
            .into()
        }
        Type::Reference(reference) => transform(
            context,
            value,
            &type_resolver::resolve(reference, context.types())?,
            position,
        )?,
        Type::Any(_) | Type::Error(_) | Type::Function(_) => {
            return Err(AnalysisError::TypeNotComparable(type_.clone()).into())
        }
    })
}

fn compile_concrete_hash_function_call(
    name: impl Into<String>,
    value: &Expression,
    type_: &Type,
    position: &Position,
) -> Expression {
    Call::new(
        Some(
            types::Function::new(
                vec![type_.clone()],
                types::Number::new(position.clone()),
                position.clone(),
            )
            .into(),
        ),
        Variable::new(name.into(), position.clone()),
        vec![value.clone()],
        position.clone(),
    )
    .into()
}

fn compile_any_function_type(position: &Position) -> types::Function {
    types::Function::new(
        vec![types::Any::new(position.clone()).into()],
        types::Number::new(position.clone()),
        position.clone(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::map_type_configuration::HASH_CONFIGURATION;
    use once_cell::sync::Lazy;
    use position::{test::PositionFake, Position};
    use pretty_assertions::assert_eq;

    static HASH_TYPE: Lazy<Type> = Lazy::new(|| types::Number::new(Position::fake()).into());

    #[test]
    fn transform_with_union() {
        let union_type = types::Union::new(
            types::Number::new(Position::fake()),
            types::None::new(Position::fake()),
            Position::fake(),
        );

        assert_eq!(
            transform(
                &CompileContext::dummy(Default::default(), Default::default()),
                &Variable::new("x", Position::fake()).into(),
                &union_type.into(),
                &Position::fake(),
            ),
            Ok(IfType::new(
                "$x",
                Variable::new("x", Position::fake()),
                vec![
                    IfTypeBranch::new(
                        types::None::new(Position::fake()),
                        Number::new(0.0, Position::fake()),
                    ),
                    IfTypeBranch::new(
                        types::Number::new(Position::fake()),
                        Call::new(
                            Some(
                                types::Function::new(
                                    vec![types::Number::new(Position::fake()).into()],
                                    HASH_TYPE.clone(),
                                    Position::fake()
                                )
                                .into()
                            ),
                            Variable::new(
                                &HASH_CONFIGURATION.number_hash_function_name,
                                Position::fake()
                            ),
                            vec![Variable::new("$x", Position::fake()).into()],
                            Position::fake()
                        ),
                    ),
                ],
                None,
                Position::fake(),
            )
            .into())
        );
    }

    #[test]
    fn transform_with_record() {
        let record_type = types::Record::new("foo", Position::fake());

        assert_eq!(
            transform(
                &CompileContext::dummy(
                    Default::default(),
                    [(
                        "foo".into(),
                        vec![types::RecordField::new(
                            "x",
                            types::None::new(Position::fake())
                        )]
                    )]
                    .into_iter()
                    .collect()
                ),
                &Variable::new("x", Position::fake()).into(),
                &record_type.clone().into(),
                &Position::fake(),
            ),
            Ok(Call::new(
                Some(
                    types::Function::new(
                        vec![record_type.clone().into()],
                        HASH_TYPE.clone(),
                        Position::fake(),
                    )
                    .into(),
                ),
                Variable::new(
                    record_type_information::compile_hash_function_name(&record_type),
                    Position::fake(),
                ),
                vec![Variable::new("x", Position::fake()).into()],
                Position::fake()
            )
            .into())
        );
    }

    #[test]
    fn fail_to_transform_with_any() {
        let any_type = types::Any::new(Position::fake());

        assert_eq!(
            transform(
                &CompileContext::dummy(Default::default(), Default::default()),
                &Variable::new("x", Position::fake()).into(),
                &any_type.clone().into(),
                &Position::fake(),
            ),
            Err(AnalysisError::TypeNotComparable(any_type.into()).into())
        );
    }

    #[test]
    fn fail_to_transform_with_function() {
        let function_type =
            types::Function::new(vec![], types::None::new(Position::fake()), Position::fake());

        assert_eq!(
            transform(
                &CompileContext::dummy(Default::default(), Default::default()),
                &Variable::new("x", Position::fake()).into(),
                &function_type.clone().into(),
                &Position::fake(),
            ),
            Err(AnalysisError::TypeNotComparable(function_type.into()).into())
        );
    }
}
