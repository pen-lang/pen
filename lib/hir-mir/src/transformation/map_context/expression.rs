use super::context_function_name;
use crate::{context::CompileContext, transformation::collection_type, CompileError};
use hir::{ir::*, types};

pub fn transform(context: &CompileContext, type_: &types::Map) -> Result<Expression, CompileError> {
    let position = type_.position();

    Ok(Call::new(
        Some(
            types::Function::new(
                vec![],
                collection_type::transform_map_context(context, position)?,
                position.clone(),
            )
            .into(),
        ),
        Variable::new(
            context_function_name(type_, context.types())?,
            position.clone(),
        ),
        vec![],
        position.clone(),
    )
    .into())
}
