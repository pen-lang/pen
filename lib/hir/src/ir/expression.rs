use super::{
    boolean::Boolean, call::Call, if_::If, if_list::IfList, if_map::IfMap, if_type::IfType,
    list::List, map::Map, none::None, number::Number, operation::Operation,
    record_construction::RecordConstruction, record_deconstruction::RecordDeconstruction,
    record_update::RecordUpdate, string::ByteString, type_coercion::TypeCoercion,
    variable::Variable, BuiltInFunction, Lambda, Let, ListComprehension, MapIterationComprehension,
    Thunk,
};
use position::Position;

#[derive(Clone, Debug, PartialEq)]
pub enum Expression {
    Boolean(Boolean),
    BuiltInFunction(BuiltInFunction),
    Call(Call),
    If(If),
    IfList(IfList),
    IfMap(IfMap),
    IfType(IfType),
    Lambda(Lambda),
    Let(Let),
    List(List),
    ListComprehension(ListComprehension),
    Map(Map),
    MapIterationComprehension(MapIterationComprehension),
    None(None),
    Number(Number),
    Operation(Operation),
    RecordConstruction(RecordConstruction),
    RecordDeconstruction(RecordDeconstruction),
    RecordUpdate(RecordUpdate),
    String(ByteString),
    Thunk(Thunk),
    TypeCoercion(TypeCoercion),
    Variable(Variable),
}

impl Expression {
    pub fn position(&self) -> &Position {
        match self {
            Self::Boolean(boolean) => boolean.position(),
            Self::BuiltInFunction(function) => function.position(),
            Self::Call(call) => call.position(),
            Self::If(if_) => if_.position(),
            Self::IfList(if_) => if_.position(),
            Self::IfMap(if_) => if_.position(),
            Self::IfType(if_) => if_.position(),
            Self::Lambda(lambda) => lambda.position(),
            Self::Let(let_) => let_.position(),
            Self::List(list) => list.position(),
            Self::ListComprehension(comprehension) => comprehension.position(),
            Self::Map(map) => map.position(),
            Self::MapIterationComprehension(comprehension) => comprehension.position(),
            Self::None(none) => none.position(),
            Self::Number(number) => number.position(),
            Self::Operation(operation) => operation.position(),
            Self::RecordConstruction(construction) => construction.position(),
            Self::RecordDeconstruction(deconstruction) => deconstruction.position(),
            Self::RecordUpdate(record_update) => record_update.position(),
            Self::String(string) => string.position(),
            Self::Thunk(thunk) => thunk.position(),
            Self::TypeCoercion(coercion) => coercion.position(),
            Self::Variable(variable) => variable.position(),
        }
    }
}

impl From<Boolean> for Expression {
    fn from(boolean: Boolean) -> Self {
        Self::Boolean(boolean)
    }
}

impl From<BuiltInFunction> for Expression {
    fn from(function: BuiltInFunction) -> Self {
        Self::BuiltInFunction(function)
    }
}

impl From<Call> for Expression {
    fn from(call: Call) -> Self {
        Self::Call(call)
    }
}

impl From<IfType> for Expression {
    fn from(if_: IfType) -> Self {
        Self::IfType(if_)
    }
}

impl From<ByteString> for Expression {
    fn from(string: ByteString) -> Self {
        Self::String(string)
    }
}

impl From<RecordConstruction> for Expression {
    fn from(construction: RecordConstruction) -> Self {
        Self::RecordConstruction(construction)
    }
}

impl From<RecordDeconstruction> for Expression {
    fn from(deconstruction: RecordDeconstruction) -> Self {
        Self::RecordDeconstruction(deconstruction)
    }
}

impl From<RecordUpdate> for Expression {
    fn from(record_update: RecordUpdate) -> Self {
        Self::RecordUpdate(record_update)
    }
}

impl From<If> for Expression {
    fn from(if_: If) -> Self {
        Self::If(if_)
    }
}

impl From<IfList> for Expression {
    fn from(if_: IfList) -> Self {
        Self::IfList(if_)
    }
}

impl From<IfMap> for Expression {
    fn from(if_: IfMap) -> Self {
        Self::IfMap(if_)
    }
}

impl From<Lambda> for Expression {
    fn from(lambda: Lambda) -> Self {
        Self::Lambda(lambda)
    }
}

impl From<Let> for Expression {
    fn from(let_: Let) -> Self {
        Self::Let(let_)
    }
}

impl From<List> for Expression {
    fn from(list: List) -> Self {
        Self::List(list)
    }
}

impl From<ListComprehension> for Expression {
    fn from(comprehension: ListComprehension) -> Self {
        Self::ListComprehension(comprehension)
    }
}

impl From<Map> for Expression {
    fn from(map: Map) -> Self {
        Self::Map(map)
    }
}

impl From<MapIterationComprehension> for Expression {
    fn from(comprehension: MapIterationComprehension) -> Self {
        Self::MapIterationComprehension(comprehension)
    }
}

impl From<None> for Expression {
    fn from(none: None) -> Self {
        Self::None(none)
    }
}

impl From<Number> for Expression {
    fn from(number: Number) -> Self {
        Self::Number(number)
    }
}

impl<T: Into<Operation>> From<T> for Expression {
    fn from(operation: T) -> Self {
        Self::Operation(operation.into())
    }
}

impl From<Thunk> for Expression {
    fn from(thunk: Thunk) -> Self {
        Self::Thunk(thunk)
    }
}

impl From<TypeCoercion> for Expression {
    fn from(coercion: TypeCoercion) -> Self {
        Self::TypeCoercion(coercion)
    }
}

impl From<Variable> for Expression {
    fn from(variable: Variable) -> Self {
        Self::Variable(variable)
    }
}
