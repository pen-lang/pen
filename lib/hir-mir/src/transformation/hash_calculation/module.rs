use super::expression;
use crate::{
    context::CompileContext,
    transformation::{collection_type, hash_calculation::function},
    CompileError,
};
use hir::{
    ir::*,
    types::{self, Type},
};

pub fn transform(context: &CompileContext, module: &Module) -> Result<Module, CompileError> {
    Ok(Module::new(
        module.type_definitions().to_vec(),
        module.type_aliases().to_vec(),
        module.foreign_declarations().to_vec(),
        module.function_declarations().to_vec(),
        module
            .function_definitions()
            .iter()
            .cloned()
            .chain(
                collection_type::collect_comparable_parameter_types(context, module)?
                    .into_iter()
                    .map(|type_| compile_function_definition(context, &type_))
                    .collect::<Result<Vec<_>, _>>()?,
            )
            .collect(),
        module.position().clone(),
    ))
}

fn compile_function_definition(
    context: &CompileContext,
    type_: &Type,
) -> Result<FunctionDefinition, CompileError> {
    const ARGUMENT_NAME: &str = "$x";

    let position = type_.position();
    let name = function::transform_name(type_, context.types())?;

    Ok(FunctionDefinition::new(
        &name,
        &name,
        Lambda::new(
            vec![Argument::new(
                ARGUMENT_NAME,
                types::Any::new(position.clone()),
            )],
            types::Number::new(position.clone()),
            IfType::new(
                ARGUMENT_NAME,
                Variable::new(ARGUMENT_NAME, position.clone()),
                vec![IfTypeBranch::new(
                    type_.clone(),
                    expression::transform(
                        context,
                        &Variable::new(ARGUMENT_NAME, position.clone()).into(),
                        type_,
                        position,
                    )?,
                )],
                None,
                position.clone(),
            ),
            position.clone(),
        ),
        None,
        false,
        position.clone(),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::compile_configuration::COMPILE_CONFIGURATION;
    use hir::test::{ModuleFake, TypeAliasFake};
    use position::{test::PositionFake, Position};

    fn transform_module(module: &Module) -> Result<Module, CompileError> {
        transform(
            &CompileContext::new(module, Some(COMPILE_CONFIGURATION.clone())),
            module,
        )
    }

    #[test]
    fn transform_comparable_type() {
        let module = Module::empty().set_type_aliases(vec![TypeAlias::fake(
            "a",
            types::Map::new(
                types::None::new(Position::fake()),
                types::None::new(Position::fake()),
                Position::fake(),
            ),
            false,
            false,
        )]);

        insta::assert_debug_snapshot!(transform_module(&module));
    }

    #[test]
    fn do_not_transform_any() {
        let module = Module::empty().set_type_aliases(vec![TypeAlias::fake(
            "a",
            types::List::new(types::Any::new(Position::fake()), Position::fake()),
            false,
            false,
        )]);

        assert_eq!(transform_module(&module), Ok(module.clone()));
    }

    #[test]
    fn do_not_transform_function() {
        let module = Module::empty().set_type_aliases(vec![TypeAlias::fake(
            "a",
            types::List::new(
                types::Function::new(vec![], types::None::new(Position::fake()), Position::fake()),
                Position::fake(),
            ),
            false,
            false,
        )]);

        assert_eq!(transform_module(&module), Ok(module.clone()));
    }
}
