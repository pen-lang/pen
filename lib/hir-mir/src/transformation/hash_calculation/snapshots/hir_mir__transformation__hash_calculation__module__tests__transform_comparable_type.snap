---
source: lib/hir-mir/src/transformation/hash_calculation/module.rs
expression: transform_module(&module)
---
Ok(
    Module {
        type_definitions: [],
        type_aliases: [
            TypeAlias {
                name: "a",
                original_name: "",
                type_: Map(
                    Map {
                        key: None(
                            None {
                                position: Position {
                                    path: "",
                                    line_number: 1,
                                    column_number: 1,
                                    line: "",
                                },
                            },
                        ),
                        value: None(
                            None {
                                position: Position {
                                    path: "",
                                    line_number: 1,
                                    column_number: 1,
                                    line: "",
                                },
                            },
                        ),
                        position: Position {
                            path: "",
                            line_number: 1,
                            column_number: 1,
                            line: "",
                        },
                    },
                ),
                public: false,
                external: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
        ],
        foreign_declarations: [],
        function_declarations: [],
        function_definitions: [
            FunctionDefinition {
                name: "hir:hash:d409c28886113c9b",
                original_name: "hir:hash:d409c28886113c9b",
                lambda: Lambda {
                    arguments: [
                        Argument {
                            name: "$x",
                            type_: Any(
                                Any {
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                        },
                    ],
                    result_type: Number(
                        Number {
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    body: IfType(
                        IfType {
                            name: "$x",
                            argument: Variable(
                                Variable {
                                    name: "$x",
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            branches: [
                                IfTypeBranch {
                                    type_: None(
                                        None {
                                            position: Position {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        },
                                    ),
                                    expression: Number(
                                        Number {
                                            value: 0.0,
                                            position: Position {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        },
                                    ),
                                },
                            ],
                            else_: None,
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    position: Position {
                        path: "",
                        line_number: 1,
                        column_number: 1,
                        line: "",
                    },
                },
                foreign_definition_configuration: None,
                public: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
        ],
        position: Position {
            path: "",
            line_number: 1,
            column_number: 1,
            line: "",
        },
    },
)
